packages/nimble-components/src/table/specs/table-column-specs/table-column-formatted-text.md

```diff
@@ -0,0 +1,241 @@
+        return value === undefined;
+    }
+}
+
+MyAppProgressColumn.registerColumn('my-app-progress-column');
+```
+
+Some of this is prototyped in the [number-column-prototype branch](https://github.com/ni/nimble/compare/main...number-column-prototype?expand=1).
+
+Other variants of this idea include:
+
+1. setting the formatting function as a property on a column element.
+2. setting an [`Intl.NumberFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) object as a property on a column element.
+
+**Pros:**
+
+-   Small memory footprint and fast data update time because formatting function is called on-demand
+-   Powerful; clients can format data however they want, including via browser APIs which are i18n-friendly
+
+**Cons:**
+
+-   Possible reduced scroll performance because formatting function is called on-demand
+-   Requires JS code to do formatting which is less convenient in frameworks like Blazor
+-   Potential cross-app inconsistency if formatting code isn't shared
+
+**Implementation Cost:**
+
+-   Expose mechanism for providing format function
+
+#### Alternative 3: Nimble provides column implementation for common use cases
+
+For common use cases we could provide column types that expose simplified formatting APIs:
+
+```html
+<nimble-table>
+    <nimble-table-column-numeric
+        field-name="progress"
+        digits-width=2
+        suffix="%"
+    >
+        Progress
+    <nimble-table-column-numeric>`
+</nimble-table>
+```
+
+**Pros:**
+
+-   Easy for clients to use since configuration is declarative
+-   Consistent formatting across apps
+
+**Cons:**
+
+-   Requires Nimble team to design simple but powerful formatting and i18n APIs
+-   Can't solve some use cases like app-specific formatting logic
+
+**Implementation Cost:**
+
+-   API design and implementation for each new column type
+
+#### Alternative 4: Client provides custom column implementation for each use case
+
+Nimble already has a mechanism for clients to provide custom columns by deriving from a base class, specifying the data fields / template / styling, and registering the column type with Nimble. We could ask clients to use this mechanism for text column types.
+
+**Pros:**
+
+-   Zero implementation cost to Nimble team
+
+**Cons:**
+
+-   Higher burden on clients to specify template, styling, etc in JS
+-   Potential for inconsistent text styling
+-   Potential cross-app inconsistency if formatting code isn't shared
+
+### Strawman Proposal
+
+For the sake of discussion my initial proposal is:
+
+1. For columns that require app-specific formatting logic I'm leaning towards "Client specifies formatting function" over "Use `table-column-text`" because it seems more like the API that app developers would expect (perhaps I'm biased by previous implementations). I'd like to do performance profiling to see how it impacts scroll performance before committing to this direction.
```

Comment:
I think the possible proposed solution outlined in the "Client specifies formatting function" section is at odds with what an app developer would expect for an API. The first variant, I think, is what clients typically expect, and resembles (I think) what [Ag-grid provides](https://www.ag-grid.com/javascript-data-grid/value-formatters/) (for example).

Reply:
I'll leave this open since I haven't addressed it (it's one level more detailed than I'm seeking agreement on at this moment, though still a good discussion to have).

I like the proposed approach of deriving from an abstract element class to define a new column type because it makes it easy to associate the formatting function with the column definition in JS code (they're in the same class). The approach of setting a formatting function via a property on the column element requires you to get a reference to the column in JS/TS/C# code which seemed more annoying to me (especially in Blazor).